libkbfs/tlf_journal.go

```diff
@@ -883,97 +889,144 @@ func (j *tlfJournal) getJournalStatus() (TLFJournalStatus, error) {
-	handle, err := MakeTlfHandle(
-		ctx, ibrmdBareHandle, j.config.usernameGetter())
-	if err != nil {
-		return TLFJournalStatus{}, err
+			irmds = append(irmds, irmd)
+		}
 	}
 
-	for _, ibrmd := range ibrmds {
-		irmd, err := j.convertImmutableBareRMDToIRMD(ctx, ibrmd, handle)
+	return irmds, nil
+}
+
+func (j *tlfJournal) getJournalStatusWithPaths(ctx context.Context,
+	cpp chainsPathPopulator) (jStatus TLFJournalStatus, err error) {
+	// This loop is limited only by the lifetime of `ctx`.
+	var unflushedPaths unflushedPathsMap
+	var complete bool
+	for {
+		var ibrmds []ImmutableBareRootMetadata
+		jStatus, ibrmds, unflushedPaths, complete, err =
+			j.getJournalStatusWithRange()
 		if err != nil {
 			return TLFJournalStatus{}, err
 		}
 
-		irmds = append(irmds, irmd)
-	}
-
-	// Make chains over the entire range to get the unflushed files.
-	// TODO: cache this chains object and update it when new revisions
-	// appear and when revisions are flushed, instead of rebuilding
-	// every time.
-	chains := newCRChainsEmpty()
-	for _, irmd := range irmds {
-		winfo := writerInfo{
-			uid:      j.uid,
-			kid:      j.key.KID(),
-			revision: irmd.Revision(),
-			// There won't be any conflicts, so no need for the
-			// username/devicename.
-		}
-		err := chains.addOps(j.config.Codec(), irmd.data, winfo,
-			irmd.localTimestamp)
-		if err != nil {
-			return TLFJournalStatus{}, nil
-		}
-	}
-	chains.mostRecentMD = irmds[len(irmds)-1]
+		// We need to make or initialize the unflushed paths.
+		if unflushedPaths == nil {
+			if !complete {
+				// Figure out the paths for the truncated MD range,
+				// but don't cache it.
+				unflushedPaths = make(unflushedPathsMap)
+				j.log.CDebugf(ctx, "Making incomplete unflushed path cache")
+				irmds, err := j.batchConvertImmutables(ctx, ibrmds)
+				if err != nil {
+					return TLFJournalStatus{}, err
+				}
+				err = addUnflushedPaths(ctx, j.uid, j.key.KID(),
+					j.config.Codec(), j.log, irmds, cpp,
+					unflushedPaths)
+				if err != nil {
+					return TLFJournalStatus{}, err
+				}
+				break
+			}
 
-	err = cpp.populateChainPaths(ctx, j.log, chains, true)
-	if err != nil {
-		return TLFJournalStatus{}, err
+			// We need to init it ourselves, or wait for someone else
+			// to do it.
+			doInit, err := j.unflushedPaths.startInitializeOrWait(ctx)
+			if err != nil {
+				return TLFJournalStatus{}, err
+			}
+
+			if doInit {
+				initSuccess := false
+				defer func() {
+					if !initSuccess || err != nil {
+						j.unflushedPaths.abortInitialization()
+					}
+				}()
+
+				j.log.CDebugf(ctx, "Initializing unflushed paths")
+				irmds, err := j.batchConvertImmutables(ctx, ibrmds)
+				if err != nil {
+					return TLFJournalStatus{}, err
+				}
+				unflushedPaths, initSuccess, err = j.unflushedPaths.initialize(
+					ctx, j.uid, j.key.KID(), j.config.Codec(), j.log, cpp,
+					irmds)
+				if err != nil {
+					return TLFJournalStatus{}, err
+				}
```

Comment:
It breaks regardless of the value of `initSuccess` -- there is a `break` below the `else` clause.  But I can re-work this to make it more clear.